client/src/components/Panels/Side/Right/Test/Test.tsx

```diff
@@ -20,7 +24,50 @@ const Test = () => {
 
   const { initialLoading, deployed, connError } = useInitialLoading();
 
-  const idl = PgProgramInfo.getProgramInfo()?.idl;
+  const idl = useMemo(() => PgProgramInfo.getProgramInfo()?.idl, []);
+  const { connection: conn } = useConnection();
+  const { currentWallet } = useCurrentWallet();
+
+  const program = useMemo(() => {
+    return idl && currentWallet
+      ? PgTest.getProgram(idl, conn, currentWallet)
+      : null;
+  }, [idl, conn, currentWallet]);
+
+  const [receivedEvents, setReceivedEvents] = useState<{
+    [eventName: string]: Object[];
+  }>({});
+
+  const handleEvent = (eventName: string, event: Object) => {
```

Comment:
Is there a reason why `Event` related code needs to live in the `Test` component instead of its own component as a child component of `Test` like the other program interactions?

Reply:
Nope, I think I did it here originally and then when I refactored thought the event listeners needed to stay here. But since there's a separate listener per event I think it's better to just put them in `ListenableEvent`. Also simplifies things because it's all in the one component so no need to map by event name any more.